<!doctype html><html lang="en" class="no-js"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><meta http-equiv="x-ua-compatible" content="ie=edge"><meta name="description" content="Compare nanopore signal to find modifications"><link rel="canonical" href="http://nanocompore.rna.rocks/demo/SampCompDB_usage/"><meta name="author" content="Adrien Leger & Tommaso Leonardi"><meta name="lang:clipboard.copy" content="Copy to clipboard"><meta name="lang:clipboard.copied" content="Copied to clipboard"><meta name="lang:search.language" content="en"><meta name="lang:search.pipeline.stopwords" content="True"><meta name="lang:search.pipeline.trimmer" content="True"><meta name="lang:search.result.none" content="No matching documents"><meta name="lang:search.result.one" content="1 matching document"><meta name="lang:search.result.other" content="# matching documents"><meta name="lang:search.tokenizer" content="[\s\-]+"><link rel="shortcut icon" href="../../assets/images/favicon.png"><meta name="generator" content="mkdocs-1.0.4, mkdocs-material-4.3.1"><title>Generate plots and tables with SampCompDB - Nanocompore</title><link rel="stylesheet" href="../../assets/stylesheets/application.4031d38b.css"><script src="../../assets/javascripts/modernizr.74668098.js"></script><link rel="stylesheet" href="../../assets/fonts/material-icons.css"></head><body dir="ltr"><svg class="md-svg"><defs><svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg></defs></svg> <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off"> <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off"> <label class="md-overlay" data-md-component="overlay" for="__drawer"></label><a href="#generate-plots-and-tables-with-sampcompdb" tabindex="1" class="md-skip">Skip to content </a><header class="md-header" data-md-component="header"><nav class="md-header-nav md-grid"><div class="md-flex"><div class="md-flex__cell md-flex__cell--shrink"><a href="http://nanocompore.rna.rocks" title="Nanocompore" class="md-header-nav__button md-logo"><i class="md-icon"></i></a></div><div class="md-flex__cell md-flex__cell--shrink"><label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label></div><div class="md-flex__cell md-flex__cell--stretch"><div class="md-flex__ellipsis md-header-nav__title" data-md-component="title"><span class="md-header-nav__topic">Nanocompore</span><span class="md-header-nav__topic">Generate plots and tables with SampCompDB</span></div></div><div class="md-flex__cell md-flex__cell--shrink"></div><div class="md-flex__cell md-flex__cell--shrink"><div class="md-header-nav__source"><a href="https://github.com/tleonardi/nanocompore/" title="Go to repository" class="md-source" data-md-source="github"><div class="md-source__icon"><svg viewBox="0 0 24 24" width="24" height="24"><use xlink:href="#__github" width="24" height="24"></use></svg></div><div class="md-source__repository">tleonardi/nanocompore</div></a></div></div></div></nav></header><div class="md-container"><main class="md-main"><div class="md-main__inner md-grid" data-md-component="container"><div class="md-sidebar md-sidebar--primary" data-md-component="navigation"><div class="md-sidebar__scrollwrap"><div class="md-sidebar__inner"><nav class="md-nav md-nav--primary" data-md-level="0"><label class="md-nav__title md-nav__title--site" for="__drawer"><a href="http://nanocompore.rna.rocks" title="Nanocompore" class="md-nav__button md-logo"><i class="md-icon"></i></a>Nanocompore</label><div class="md-nav__source"><a href="https://github.com/tleonardi/nanocompore/" title="Go to repository" class="md-source" data-md-source="github"><div class="md-source__icon"><svg viewBox="0 0 24 24" width="24" height="24"><use xlink:href="#__github" width="24" height="24"></use></svg></div><div class="md-source__repository">tleonardi/nanocompore</div></a></div><ul class="md-nav__list" data-md-scrollfix><li class="md-nav__item"><a href="../.." title="Home" class="md-nav__link">Home</a></li><li class="md-nav__item"><a href="../../installation/" title="Installation instructions" class="md-nav__link">Installation instructions</a></li><li class="md-nav__item"><a href="../../data_preparation/" title="Data preparation" class="md-nav__link">Data preparation</a></li><li class="md-nav__item md-nav__item--active md-nav__item--nested"><input class="md-toggle md-nav__toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" checked><label class="md-nav__link" for="nav-4">Usage</label><nav class="md-nav" data-md-component="collapsible" data-md-level="1"><label class="md-nav__title" for="nav-4">Usage</label><ul class="md-nav__list" data-md-scrollfix><li class="md-nav__item"><a href="../../usage/" title="How to use Nanocompore" class="md-nav__link">How to use Nanocompore</a></li><li class="md-nav__item"><a href="../SampComp_usage/" title="Compare samples with SampComp" class="md-nav__link">Compare samples with SampComp</a></li><li class="md-nav__item md-nav__item--active"><input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc"><label class="md-nav__link md-nav__link--active" for="__toc">Generate plots and tables with SampCompDB</label><a href="./" title="Generate plots and tables with SampCompDB" class="md-nav__link md-nav__link--active">Generate plots and tables with SampCompDB</a><nav class="md-nav md-nav--secondary"><label class="md-nav__title" for="__toc">Table of contents</label><ul class="md-nav__list" data-md-scrollfix><li class="md-nav__item"><a href="#import-the-package" title="Import the package" class="md-nav__link">Import the package</a></li><li class="md-nav__item"><a href="#load-the-database-with-sampcompdb" title="Load the database with SampCompDB" class="md-nav__link">Load the database with SampCompDB</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#basic-initialisation" title="Basic initialisation" class="md-nav__link">Basic initialisation</a></li></ul></nav></li><li class="md-nav__item"><a href="#generate-text-reports" title="Generate text reports" class="md-nav__link">Generate text reports</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#save_report" title="save_report" class="md-nav__link">save_report</a></li><li class="md-nav__item"><a href="#save_shift_stats" title="save_shift_stats" class="md-nav__link">save_shift_stats</a></li><li class="md-nav__item"><a href="#save_to_bed" title="save_to_bed" class="md-nav__link">save_to_bed</a></li></ul></nav></li><li class="md-nav__item"><a href="#generate-plots" title="Generate plots" class="md-nav__link">Generate plots</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#extra-imports-for-the-plotting-library" title="Extra imports for the plotting library" class="md-nav__link">Extra imports for the plotting library</a></li><li class="md-nav__item"><a href="#plot_pvalue" title="plot_pvalue" class="md-nav__link">plot_pvalue</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#examples-from-simulated-dataset" title="Examples from simulated dataset" class="md-nav__link">Examples from simulated dataset</a></li><li class="md-nav__item"><a href="#example-from-real-yeast-dataset-with-extended-sequence-context" title="Example from real yeast dataset with extended sequence context" class="md-nav__link">Example from real yeast dataset with extended sequence context</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_signal" title="plot_signal" class="md-nav__link">plot_signal</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#examples-from-simulated-dataset_1" title="Examples from simulated dataset" class="md-nav__link">Examples from simulated dataset</a></li><li class="md-nav__item"><a href="#example-from-real-yeast-dataset" title="Example from real yeast dataset" class="md-nav__link">Example from real yeast dataset</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_coverage" title="plot_coverage" class="md-nav__link">plot_coverage</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#example-from-real-yeast-dataset_1" title="Example from real yeast dataset" class="md-nav__link">Example from real yeast dataset</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_kmers_stats" title="plot_kmers_stats" class="md-nav__link">plot_kmers_stats</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#example-from-real-yeast-dataset_2" title="Example from real yeast dataset" class="md-nav__link">Example from real yeast dataset</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_position" title="plot_position" class="md-nav__link">plot_position</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#example-from-simulated-dataset" title="Example from simulated dataset" class="md-nav__link">Example from simulated dataset</a></li></ul></nav></li></ul></nav></li></ul></nav></li><li class="md-nav__item"><a href="../SimReads_usage/" title="Generate simulated reads" class="md-nav__link">Generate simulated reads</a></li></ul></nav></li><li class="md-nav__item"><a href="../../alternative/" title="Alternative and complementary packages" class="md-nav__link">Alternative and complementary packages</a></li><li class="md-nav__item"><a href="../../citing/" title="Citing" class="md-nav__link">Citing</a></li><li class="md-nav__item"><a href="../../contributing/" title="Contributing" class="md-nav__link">Contributing</a></li><li class="md-nav__item"><a href="../../code_of_conduct/" title="Code of conduct" class="md-nav__link">Code of conduct</a></li><li class="md-nav__item"><a href="../../licence/" title="Licence" class="md-nav__link">Licence</a></li><li class="md-nav__item"><a href="../../changelog/" title="Changelog" class="md-nav__link">Changelog</a></li></ul></nav></div></div></div><div class="md-sidebar md-sidebar--secondary" data-md-component="toc"><div class="md-sidebar__scrollwrap"><div class="md-sidebar__inner"><nav class="md-nav md-nav--secondary"><label class="md-nav__title" for="__toc">Table of contents</label><ul class="md-nav__list" data-md-scrollfix><li class="md-nav__item"><a href="#import-the-package" title="Import the package" class="md-nav__link">Import the package</a></li><li class="md-nav__item"><a href="#load-the-database-with-sampcompdb" title="Load the database with SampCompDB" class="md-nav__link">Load the database with SampCompDB</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#basic-initialisation" title="Basic initialisation" class="md-nav__link">Basic initialisation</a></li></ul></nav></li><li class="md-nav__item"><a href="#generate-text-reports" title="Generate text reports" class="md-nav__link">Generate text reports</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#save_report" title="save_report" class="md-nav__link">save_report</a></li><li class="md-nav__item"><a href="#save_shift_stats" title="save_shift_stats" class="md-nav__link">save_shift_stats</a></li><li class="md-nav__item"><a href="#save_to_bed" title="save_to_bed" class="md-nav__link">save_to_bed</a></li></ul></nav></li><li class="md-nav__item"><a href="#generate-plots" title="Generate plots" class="md-nav__link">Generate plots</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#extra-imports-for-the-plotting-library" title="Extra imports for the plotting library" class="md-nav__link">Extra imports for the plotting library</a></li><li class="md-nav__item"><a href="#plot_pvalue" title="plot_pvalue" class="md-nav__link">plot_pvalue</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#examples-from-simulated-dataset" title="Examples from simulated dataset" class="md-nav__link">Examples from simulated dataset</a></li><li class="md-nav__item"><a href="#example-from-real-yeast-dataset-with-extended-sequence-context" title="Example from real yeast dataset with extended sequence context" class="md-nav__link">Example from real yeast dataset with extended sequence context</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_signal" title="plot_signal" class="md-nav__link">plot_signal</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#examples-from-simulated-dataset_1" title="Examples from simulated dataset" class="md-nav__link">Examples from simulated dataset</a></li><li class="md-nav__item"><a href="#example-from-real-yeast-dataset" title="Example from real yeast dataset" class="md-nav__link">Example from real yeast dataset</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_coverage" title="plot_coverage" class="md-nav__link">plot_coverage</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#example-from-real-yeast-dataset_1" title="Example from real yeast dataset" class="md-nav__link">Example from real yeast dataset</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_kmers_stats" title="plot_kmers_stats" class="md-nav__link">plot_kmers_stats</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#example-from-real-yeast-dataset_2" title="Example from real yeast dataset" class="md-nav__link">Example from real yeast dataset</a></li></ul></nav></li><li class="md-nav__item"><a href="#plot_position" title="plot_position" class="md-nav__link">plot_position</a><nav class="md-nav"><ul class="md-nav__list"><li class="md-nav__item"><a href="#example-from-simulated-dataset" title="Example from simulated dataset" class="md-nav__link">Example from simulated dataset</a></li></ul></nav></li></ul></nav></li></ul></nav></div></div></div><div class="md-content"><article class="md-content__inner md-typeset"><a href="https://github.com/tleonardi/nanocompore/edit/master/docs/demo/SampCompDB_usage.ipynb" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a><h1 id="generate-plots-and-tables-with-sampcompdb">Generate plots and tables with SampCompDB</h1>
<p><code class="codehilite"><span class="n">SampComp</span></code> creates a python object database (shelve DBM) containing the statistical analysis results. The API directly returns a <code class="codehilite"><span class="n">SampCompDB</span></code> object wrapping the shelve DB. It is also possible to reload the <code class="codehilite"><span class="n">SampCompDB</span></code> latter using the db file path prefix. <code class="codehilite"><span class="n">SampCompDB</span></code> also need a FASTA file to get the corresponding reference id sequence and accept an optional BED file containing genomic annotations. SampCompDB provide a large selection of simple high level function to plot and export the results.</p>
<p>At the moment <code class="codehilite"><span class="n">SampCompDB</span></code> is only accessible through the python API.</p>
<h2 id="import-the-package">Import the package</h2>
<div class="codehilite"><pre><span></span><span class="kn">from</span> <span class="nn">nanocompore.SampCompDB</span> <span class="kn">import</span> <span class="n">SampCompDB</span><span class="p">,</span> <span class="n">jhelp</span>
</pre></div>

<h2 id="load-the-database-with-sampcompdb">Load the database with SampCompDB</h2>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span> <span class="p">(</span><span class="n">SampCompDB</span><span class="p">)</span>
</pre></div>

<p><strong>SampCompDB</strong> (db_fn, fasta_fn, bed_fn, run_type, log_level)</p>
<p>Wrapper over the result shelve SampComp </p>
<hr />
<ul>
<li><strong>db_fn</strong> (required) [str]</li>
</ul>
<p>Path to a database file previously created with SampComp</p>
<ul>
<li><strong>fasta_fn</strong> (required) [str]</li>
</ul>
<p>Path to a fasta file corresponding to the reference used for read alignemnt</p>
<ul>
<li><strong>bed_fn</strong> (default: None) [str]</li>
</ul>
<p>Path to a BED file containing the annotation of the transcriptome used as reference when mapping</p>
<ul>
<li><strong>run_type</strong> (default: RNA) [str]</li>
</ul>
<p>Define the run type model to import {RNA, DNA}</p>
<ul>
<li><strong>log_level</strong> (default: info) [str]</li>
</ul>
<p>Set the log level. {warning,info,debug}"</p>
<h3 id="basic-initialisation">Basic initialisation</h3>
<div class="codehilite"><pre><span></span><span class="c1"># Load database</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span>
    <span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span>
    <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">)</span>

<span class="c1"># Print general metadata information</span>
<span class="k">print</span> <span class="p">(</span><span class="n">db</span><span class="p">)</span>

<span class="c1"># Prit list of references containing valid data</span>
<span class="k">print</span> <span class="p">(</span><span class="n">db</span><span class="o">.</span><span class="n">ref_id_list</span><span class="p">)</span>
</pre></div>

<div class="codehilite"><pre><span></span><span class="n">Loading</span><span class="w"> </span><span class="n">SampCompDB</span><span class="w"></span>
<span class="n">Calculate</span><span class="w"> </span><span class="n">results</span><span class="w"></span>


<span class="o">[</span><span class="n">SampCompDB</span><span class="o">]</span><span class="w"></span>
<span class="w">    </span><span class="nl">package_name</span><span class="p">:</span><span class="w"> </span><span class="n">nanocompore</span><span class="w"></span>
<span class="w">    </span><span class="nl">package_version</span><span class="p">:</span><span class="w"> </span><span class="mf">1.0.0</span><span class="n">rc3</span><span class="o">-</span><span class="n">dev</span><span class="w"></span>
<span class="w">    </span><span class="nc">timestamp</span><span class="err">:</span><span class="w"> </span><span class="mi">2019</span><span class="o">-</span><span class="mi">06</span><span class="o">-</span><span class="mi">19</span><span class="w"> </span><span class="mi">14</span><span class="err">:</span><span class="mi">11</span><span class="err">:</span><span class="mf">42.582069</span><span class="w"></span>
<span class="w">    </span><span class="nl">comparison_methods</span><span class="p">:</span><span class="w"> </span><span class="o">[</span><span class="n">&#39;GMM&#39;, &#39;KS&#39;</span><span class="o">]</span><span class="w"></span>
<span class="w">    </span><span class="nl">pvalue_tests</span><span class="p">:</span><span class="w"> </span><span class="o">[</span><span class="n">&#39;GMM_anova_pvalue&#39;, &#39;KS_dwell_pvalue&#39;, &#39;KS_intensity_pvalue&#39;</span><span class="o">]</span><span class="w"></span>
<span class="w">    </span><span class="nl">sequence_context</span><span class="p">:</span><span class="w"> </span><span class="mi">0</span><span class="w"></span>
<span class="w">    </span><span class="nl">min_coverage</span><span class="p">:</span><span class="w"> </span><span class="mi">30</span><span class="w"></span>
<span class="w">    </span><span class="nl">n_samples</span><span class="p">:</span><span class="w"> </span><span class="mi">4</span><span class="w"></span>
<span class="w">    </span><span class="n">Number</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="k">references</span><span class="err">:</span><span class="w"> </span><span class="mi">5</span><span class="w"></span>

<span class="o">[</span><span class="n">&#39;ref_0002&#39;, &#39;ref_0000&#39;, &#39;ref_0003&#39;, &#39;ref_0001&#39;, &#39;ref_0004&#39;</span><span class="o">]</span><span class="w"></span>
</pre></div>


<h2 id="generate-text-reports">Generate text reports</h2>
<p><code class="codehilite"><span class="n">SampCompDB</span></code> can generate 3 types of text reports:
* Tabulated statistics =&gt; <strong><code class="codehilite"><span class="n">save_report</span></code></strong>
* Tabulated intensity and dwell values per conditions =&gt; <strong><code class="codehilite"><span class="n">save_shift_stats</span></code></strong>
* BED significant genomic positions =&gt;  <strong><code class="codehilite"><span class="n">save_to_bed</span></code></strong></p>
<p>In addition, we also conveniently wrapped all 3 methods in <strong><code class="codehilite"><span class="n">save_all</span></code></strong>.</p>
<h3 id="save_report">save_report</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">save_report</span><span class="p">)</span>
</pre></div>

<p><strong>save_report</strong> (output_fn)</p>
<p>Saves a tabulated text dump of the database containing all the statistical results for all the positions</p>
<hr />
<ul>
<li><strong>output_fn</strong> (default: None) [str]</li>
</ul>
<p>Path to file where to write the data. If None, data is returned to the standard output.</p>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>

<span class="c1"># Save report</span>
<span class="n">db</span><span class="o">.</span><span class="n">save_report</span> <span class="p">(</span><span class="n">output_fn</span><span class="o">=</span><span class="s2">&quot;./results/simulated_report.tsv&quot;</span><span class="p">)</span>

<span class="c1"># Visualise first lines</span>
<span class="err">!</span><span class="n">head</span> <span class="s2">&quot;./results/simulated_report.tsv&quot;</span>
</pre></div>

<div class="codehilite"><pre><span></span><span class="n">pos</span> <span class="n">chr</span> <span class="n">genomicPos</span>  <span class="n">ref_id</span>  <span class="n">strand</span>  <span class="n">ref_kmer</span>    <span class="n">GMM_anova_pvalue</span>    <span class="n">KS_dwell_pvalue</span> <span class="n">KS_intensity_pvalue</span> <span class="n">GMM_cov_type</span>    <span class="n">GMM_n_clust</span> <span class="n">cluster_counts</span>  <span class="n">Anova_delta_logit</span>
<span class="mi">0</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">AAGGA</span>   <span class="mi">0</span><span class="p">.</span><span class="mi">010390555392639545</span>    <span class="mi">3</span><span class="p">.</span><span class="mi">7826176625121133</span><span class="n">e</span><span class="o">-</span><span class="mi">28</span>  <span class="mi">7</span><span class="p">.</span><span class="mi">372203704353806</span><span class="n">e</span><span class="o">-</span><span class="mi">13</span>   <span class="k">full</span>    <span class="mi">2</span>   <span class="n">Modified_rep1</span><span class="p">:</span><span class="mi">4</span><span class="o">/</span><span class="mi">51</span><span class="n">__Modified_rep2</span><span class="p">:</span><span class="mi">4</span><span class="o">/</span><span class="mi">56</span><span class="n">__Unmodified_rep1</span><span class="p">:</span><span class="mi">39</span><span class="o">/</span><span class="mi">16</span><span class="n">__Unmodified_rep2</span><span class="p">:</span><span class="mi">49</span><span class="o">/</span><span class="mi">11</span>    <span class="o">-</span><span class="mi">3</span><span class="p">.</span><span class="mi">5291008135000004</span>
<span class="mi">1</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">AGGAC</span>   <span class="mi">0</span><span class="p">.</span><span class="mi">018531016868604602</span>    <span class="mi">5</span><span class="p">.</span><span class="mi">021033530354203</span><span class="n">e</span><span class="o">-</span><span class="mi">38</span>   <span class="mi">2</span><span class="p">.</span><span class="mi">6192924303287954</span><span class="n">e</span><span class="o">-</span><span class="mi">32</span>  <span class="k">full</span>    <span class="mi">2</span>   <span class="n">Modified_rep1</span><span class="p">:</span><span class="mi">49</span><span class="o">/</span><span class="mi">6</span><span class="n">__Modified_rep2</span><span class="p">:</span><span class="mi">54</span><span class="o">/</span><span class="mi">6</span><span class="n">__Unmodified_rep1</span><span class="p">:</span><span class="mi">0</span><span class="o">/</span><span class="mi">55</span><span class="n">__Unmodified_rep2</span><span class="p">:</span><span class="mi">3</span><span class="o">/</span><span class="mi">57</span>  <span class="mi">5</span><span class="p">.</span><span class="mi">363518116</span>
<span class="mi">2</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">GGACT</span>   <span class="mi">0</span><span class="p">.</span><span class="mi">016650130591381845</span>    <span class="mi">9</span><span class="p">.</span><span class="mi">083535018319583</span><span class="n">e</span><span class="o">-</span><span class="mi">23</span>   <span class="mi">1</span><span class="p">.</span><span class="mi">9347009060981327</span><span class="n">e</span><span class="o">-</span><span class="mi">19</span>  <span class="k">full</span>    <span class="mi">2</span>   <span class="n">Modified_rep1</span><span class="p">:</span><span class="mi">48</span><span class="o">/</span><span class="mi">7</span><span class="n">__Modified_rep2</span><span class="p">:</span><span class="mi">53</span><span class="o">/</span><span class="mi">7</span><span class="n">__Unmodified_rep1</span><span class="p">:</span><span class="mi">13</span><span class="o">/</span><span class="mi">42</span><span class="n">__Unmodified_rep2</span><span class="p">:</span><span class="mi">7</span><span class="o">/</span><span class="mi">53</span> <span class="mi">3</span><span class="p">.</span><span class="mi">3768032759999995</span>
<span class="mi">3</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">GACTG</span>   <span class="n">nan</span> <span class="mi">1</span><span class="p">.</span><span class="mi">5167780008574906</span><span class="n">e</span><span class="o">-</span><span class="mi">20</span>  <span class="mi">5</span><span class="p">.</span><span class="mi">038173683218352</span><span class="n">e</span><span class="o">-</span><span class="mi">10</span>   <span class="k">full</span>    <span class="mi">1</span>   <span class="n">NC</span>  <span class="n">nan</span>
<span class="mi">4</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">ACTGC</span>   <span class="n">nan</span> <span class="mi">0</span><span class="p">.</span><span class="mi">7103128384794648</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">2143969244232542</span>  <span class="k">full</span>    <span class="mi">1</span>   <span class="n">NC</span>  <span class="n">nan</span>
<span class="mi">5</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">CTGCG</span>   <span class="n">nan</span> <span class="mi">0</span><span class="p">.</span><span class="mi">5427546298577994</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">9441367354284554</span>  <span class="k">full</span>    <span class="mi">1</span>   <span class="n">NC</span>  <span class="n">nan</span>
<span class="mi">6</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">TGCGC</span>   <span class="n">nan</span> <span class="mi">0</span><span class="p">.</span><span class="mi">6283003014678024</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">2835540266938749</span>  <span class="k">full</span>    <span class="mi">1</span>   <span class="n">NC</span>  <span class="n">nan</span>
<span class="mi">7</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">GCGCC</span>   <span class="n">nan</span> <span class="mi">0</span><span class="p">.</span><span class="mi">7818041839421846</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">2143969244232542</span>  <span class="k">full</span>    <span class="mi">1</span>   <span class="n">NC</span>  <span class="n">nan</span>
<span class="mi">8</span>   <span class="n">NA</span>  <span class="n">NA</span>  <span class="n">ref_0002</span>    <span class="n">NA</span>  <span class="n">CGCCG</span>   <span class="n">nan</span> <span class="mi">0</span><span class="p">.</span><span class="mi">4544427448826311</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">8445913073368677</span>  <span class="k">full</span>    <span class="mi">1</span>   <span class="n">NC</span>  <span class="n">nan</span>
</pre></div>


<h3 id="save_shift_stats">save_shift_stats</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">save_shift_stats</span><span class="p">)</span>
</pre></div>

<p><strong>save_shift_stats</strong> (output_fn)</p>
<p>Save the mean, median and sd intensity and dwell time for each condition and for each position.
This can be used to evaluate the intensity of the shift for significant positions.</p>
<hr />
<ul>
<li><strong>output_fn</strong> (default: None)</li>
</ul>
<p>Path to file where to write the data. If None, data is returned to the standard output.</p>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>

<span class="c1"># Save report</span>
<span class="n">db</span><span class="o">.</span><span class="n">save_shift_stats</span> <span class="p">(</span><span class="n">output_fn</span><span class="o">=</span><span class="s2">&quot;./results/simulated_shift.tsv&quot;</span><span class="p">)</span>

<span class="c1"># Visualise first lines</span>
<span class="err">!</span><span class="n">head</span> <span class="s2">&quot;./results/simulated_shift.tsv&quot;</span>
</pre></div>

<div class="codehilite"><pre><span></span><span class="n">red_if</span>  <span class="n">pos</span> <span class="n">c1_mean_intensity</span>   <span class="n">c2_mean_intensity</span>   <span class="n">c1_median_intensity</span> <span class="n">c2_median_intensity</span> <span class="n">c1_sd_intensity</span> <span class="n">c2_sd_intensity</span> <span class="n">c1_mean_dwell</span>   <span class="n">c2_mean_dwell</span>   <span class="n">c1_median_dwell</span> <span class="n">c2_median_dwell</span> <span class="n">c1_sd_dwell</span> <span class="n">c2_sd_dwell</span>
<span class="n">ref_0001</span>    <span class="mi">0</span>   <span class="mi">99</span><span class="p">.</span><span class="mi">31924275131445</span>   <span class="mi">95</span><span class="p">.</span><span class="mi">20865205128713</span>   <span class="mi">99</span><span class="p">.</span><span class="mi">30450652082747</span>   <span class="mi">95</span><span class="p">.</span><span class="mi">48506067828346</span>   <span class="mi">3</span><span class="p">.</span><span class="mi">4542546529463602</span>  <span class="mi">3</span><span class="p">.</span><span class="mi">2457394860292355</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">02892371350987724</span> <span class="mi">0</span><span class="p">.</span><span class="mi">011492822977360408</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">026110578053659444</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">008409393806440126</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">012309163786594407</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">010097740816656625</span>
<span class="n">ref_0001</span>    <span class="mi">1</span>   <span class="mi">95</span><span class="p">.</span><span class="mi">8472415436861</span>    <span class="mi">86</span><span class="p">.</span><span class="mi">76320109231014</span>   <span class="mi">96</span><span class="p">.</span><span class="mi">51658647295791</span>   <span class="mi">86</span><span class="p">.</span><span class="mi">5897261563199</span>    <span class="mi">4</span><span class="p">.</span><span class="mi">818212253968896</span>   <span class="mi">3</span><span class="p">.</span><span class="mi">462616900325914</span>   <span class="mi">0</span><span class="p">.</span><span class="mi">049904028745349535</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">014284063287085003</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">049430992173562574</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">00838888597355578</span> <span class="mi">0</span><span class="p">.</span><span class="mi">01891436186841055</span> <span class="mi">0</span><span class="p">.</span><span class="mi">01469932192592977</span>
<span class="n">ref_0001</span>    <span class="mi">2</span>   <span class="mi">99</span><span class="p">.</span><span class="mi">95956311201668</span>   <span class="mi">91</span><span class="p">.</span><span class="mi">94076655256666</span>   <span class="mi">99</span><span class="p">.</span><span class="mi">2359574515384</span>    <span class="mi">91</span><span class="p">.</span><span class="mi">66515374416255</span>   <span class="mi">6</span><span class="p">.</span><span class="mi">732308978957026</span>   <span class="mi">5</span><span class="p">.</span><span class="mi">022072514781664</span>   <span class="mi">0</span><span class="p">.</span><span class="mi">027605907187015038</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">013686251128416578</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">024211963436724607</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">009798007086063236</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">014206789096974112</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">01220646673010648</span>
<span class="n">ref_0001</span>    <span class="mi">3</span>   <span class="mi">123</span><span class="p">.</span><span class="mi">38251652061109</span>  <span class="mi">120</span><span class="p">.</span><span class="mi">39646776876853</span>  <span class="mi">123</span><span class="p">.</span><span class="mi">6421269382127</span>   <span class="mi">120</span><span class="p">.</span><span class="mi">37875855935422</span>  <span class="mi">3</span><span class="p">.</span><span class="mi">595296347867109</span>   <span class="mi">3</span><span class="p">.</span><span class="mi">44338463383404</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">028760344887274834</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">013945564031975161</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">02351276462197541</span> <span class="mi">0</span><span class="p">.</span><span class="mi">010354002459941074</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">017565155869793847</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">011016717175902649</span>
<span class="n">ref_0001</span>    <span class="mi">4</span>   <span class="mi">126</span><span class="p">.</span><span class="mi">94163302104937</span>  <span class="mi">128</span><span class="p">.</span><span class="mi">10426869865935</span>  <span class="mi">126</span><span class="p">.</span><span class="mi">55839976875221</span>  <span class="mi">127</span><span class="p">.</span><span class="mi">8992550363245</span>   <span class="mi">3</span><span class="p">.</span><span class="mi">673183364723787</span>   <span class="mi">4</span><span class="p">.</span><span class="mi">450043557112173</span>   <span class="mi">0</span><span class="p">.</span><span class="mi">011798347544928565</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">013913244140038266</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">008085339548882277</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">009808706553017481</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">010235524821933415</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">012247487588286306</span>
<span class="n">ref_0001</span>    <span class="mi">5</span>   <span class="mi">67</span><span class="p">.</span><span class="mi">79566862215866</span>   <span class="mi">67</span><span class="p">.</span><span class="mi">87066906855172</span>   <span class="mi">67</span><span class="p">.</span><span class="mi">57337199388174</span>   <span class="mi">67</span><span class="p">.</span><span class="mi">66864765166001</span>   <span class="mi">2</span><span class="p">.</span><span class="mi">676099462286074</span>   <span class="mi">2</span><span class="p">.</span><span class="mi">3271061496852905</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">011811464983494136</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">012975245722562987</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">007659091594070025</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">007913024971489328</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">01114312094701346</span> <span class="mi">0</span><span class="p">.</span><span class="mi">013957539542755816</span>
<span class="n">ref_0001</span>    <span class="mi">6</span>   <span class="mi">66</span><span class="p">.</span><span class="mi">155214143451</span> <span class="mi">65</span><span class="p">.</span><span class="mi">87786841257429</span>   <span class="mi">66</span><span class="p">.</span><span class="mi">01760504576498</span>   <span class="mi">65</span><span class="p">.</span><span class="mi">80903439895116</span>   <span class="mi">1</span><span class="p">.</span><span class="mi">5942871182383656</span>  <span class="mi">1</span><span class="p">.</span><span class="mi">9921259312963864</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">013708376752126073</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">012800371955813343</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">009551895684442788</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">008605404284103613</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">014921164088938836</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">011530961046002914</span>
<span class="n">ref_0001</span>    <span class="mi">7</span>   <span class="mi">69</span><span class="p">.</span><span class="mi">61169482565616</span>   <span class="mi">68</span><span class="p">.</span><span class="mi">99766496300363</span>   <span class="mi">69</span><span class="p">.</span><span class="mi">41928437670197</span>   <span class="mi">68</span><span class="p">.</span><span class="mi">95292840911311</span>   <span class="mi">2</span><span class="p">.</span><span class="mi">936936891666408</span>   <span class="mi">2</span><span class="p">.</span><span class="mi">575355539206547</span>   <span class="mi">0</span><span class="p">.</span><span class="mi">011149410753827556</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">011084327064933388</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">008581162093140242</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">007464155848405434</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">009824895741362088</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">01064339382049097</span>
<span class="n">ref_0001</span>    <span class="mi">8</span>   <span class="mi">82</span><span class="p">.</span><span class="mi">34611997642149</span>   <span class="mi">82</span><span class="p">.</span><span class="mi">05089836082209</span>   <span class="mi">82</span><span class="p">.</span><span class="mi">24692305018203</span>   <span class="mi">81</span><span class="p">.</span><span class="mi">9154038674852</span>    <span class="mi">2</span><span class="p">.</span><span class="mi">5845230611008763</span>  <span class="mi">2</span><span class="p">.</span><span class="mi">4502360451089737</span>  <span class="mi">0</span><span class="p">.</span><span class="mi">01267473742329643</span> <span class="mi">0</span><span class="p">.</span><span class="mi">012320988335592328</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">010373590682259867</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">009046996378582178</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">009070752556799426</span>    <span class="mi">0</span><span class="p">.</span><span class="mi">00959451914969272</span>
</pre></div>


<h3 id="save_to_bed">save_to_bed</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">save_to_bed</span><span class="p">)</span>
</pre></div>

<p><strong>save_to_bed</strong> (output_fn, bedgraph, pvalue_field, pvalue_thr, span, convert, assembly, title)</p>
<p>Save the position of significant positions in the genome space in BED6 or BEDGRAPH format.
The resulting file can be used in a genome browser to visualise significant genomic locations.
The option is only available if <code class="codehilite"><span class="n">SampCompDB</span></code> if initialised with a BED file containing genome annotations.</p>
<hr />
<ul>
<li><strong>output_fn</strong> (default: None)</li>
</ul>
<p>Path to file where to write the data</p>
<ul>
<li><strong>bedgraph</strong> (default: False)</li>
</ul>
<p>save file in bedgraph format instead of bed</p>
<ul>
<li><strong>pvalue_field</strong> (default: None)</li>
</ul>
<p>specifies what column to use as BED score (field 5, as -log10)</p>
<ul>
<li><strong>pvalue_thr</strong> (default: 0.01)</li>
</ul>
<p>only report positions with pvalue&lt;=thr</p>
<ul>
<li><strong>span</strong> (default: 5)</li>
</ul>
<p>The size of each BED feature. If size=5 (default) features correspond to kmers. If size=1 features correspond to the first base of each kmer.</p>
<ul>
<li><strong>convert</strong> (default: None)</li>
</ul>
<p>one of 'ensembl_to_ucsc' or 'ucsc_to_ensembl". Convert chromosome named between Ensembl and Ucsc conventions</p>
<ul>
<li><strong>assembly</strong> (default: None)</li>
</ul>
<p>required if convert is used. One of "hg38" or "mm10"</p>
<ul>
<li><strong>title</strong> (default: None)</li>
</ul>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">bed_fn</span><span class="o">=</span><span class="s2">&quot;references/simulated/annot.bed&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>

<span class="c1"># Save report</span>
<span class="n">db</span><span class="o">.</span><span class="n">save_to_bed</span> <span class="p">(</span><span class="n">output_fn</span><span class="o">=</span><span class="s2">&quot;./results/simulated_sig_positions.bed&quot;</span><span class="p">)</span>

<span class="c1"># Visualise first lines</span>
<span class="err">!</span><span class="n">head</span> <span class="s2">&quot;./results/simulated_sig_positions.bed&quot;</span>
</pre></div>

<h2 id="generate-plots">Generate plots</h2>
<p><code class="codehilite"><span class="n">SampCompDB</span></code> comes with a range of methods to visualise the data and explore candidates.</p>
<ul>
<li><strong><code class="codehilite"><span class="n">plot_pvalue</span></code></strong>: Plot the <code class="codehilite"><span class="o">-</span><span class="n">log</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span></code> of the pvalues obtained for all the statistical methods at reference level</li>
<li><strong><code class="codehilite"><span class="n">plot_signal</span></code></strong>: Generate comparative plots of both median intensity and dwell time for each condition at read level </li>
<li><strong><code class="codehilite"><span class="n">plot_coverage</span></code></strong>: Plot the read coverage over a reference for all samples analysed </li>
<li><strong><code class="codehilite"><span class="n">plot_kmers_stats</span></code></strong>: Fancy version of <code class="codehilite"><span class="n">plot_coverage</span></code> that also report missing, mismatching and undefined kmers status from Nanopolish </li>
<li><strong><code class="codehilite"><span class="n">plot_position</span></code></strong>: Allow to visualise the distribution of intensity and dwell time in 2D for a single position</li>
</ul>
<h3 id="extra-imports-for-the-plotting-library">Extra imports for the plotting library</h3>
<p>Matplotlib is required to use the ploting methods in Jupyter </p>
<div class="codehilite"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">pl</span>
<span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span>
</pre></div>

<h3 id="plot_pvalue">plot_pvalue</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">plot_pvalue</span><span class="p">)</span>
</pre></div>

<p><strong>plot_pvalue</strong> (ref_id, start, end, kind, threshold, figsize, palette, plot_style, tests)</p>
<p>Plot pvalues per position (by default plot all fields starting by "pvalue")</p>
<hr />
<ul>
<li><strong>ref_id</strong> (required) [str]</li>
</ul>
<p>Valid reference id name in the database</p>
<ul>
<li><strong>start</strong> (default: None) [int]</li>
</ul>
<p>Start coordinate</p>
<ul>
<li><strong>end</strong> (default: None) [int]</li>
</ul>
<p>End coordinate (included)</p>
<ul>
<li><strong>kind</strong> (default: lineplot) [str]</li>
</ul>
<p>kind of plot to represent the data. {lineplot,barplot}</p>
<ul>
<li>
<p><strong>threshold</strong> (default: 0.01) [float]</p>
</li>
<li>
<p><strong>figsize</strong> (default: (30, 10)) [tuple]</p>
</li>
</ul>
<p>Length and heigh of the output plot</p>
<ul>
<li><strong>palette</strong> (default: Set2) [str]</li>
</ul>
<p>Colormap. See <a href="https://matplotlib.org/users/colormaps.html">https://matplotlib.org/users/colormaps.html</a>, <a href="https://matplotlib.org/examples/color/named_colors.html">https://matplotlib.org/examples/color/named_colors.html</a></p>
<ul>
<li><strong>plot_style</strong> (default: ggplot) [str]</li>
</ul>
<p>Matplotlib plotting style. See <a href="https://matplotlib.org/users/style_sheets.html">https://matplotlib.org/users/style_sheets.html</a></p>
<ul>
<li><strong>tests</strong> (default: None) [str]</li>
</ul>
<p>Limit the pvalue methods shown in the plot. Either a list of methods or a string coresponding to a part of the name</p>
<h4 id="examples-from-simulated-dataset">Examples from simulated dataset</h4>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_pvalue</span> <span class="p">(</span><span class="s2">&quot;ref_0000&quot;</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_27_0.png" /></p>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_stats_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_pvalue</span> <span class="p">(</span><span class="s2">&quot;ref_0001&quot;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set1&quot;</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_28_0.png" /></p>
<h4 id="example-from-real-yeast-dataset-with-extended-sequence-context">Example from real yeast dataset with extended sequence context</h4>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/yeast_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/yeast/Yeast_transcriptome.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_pvalue</span> <span class="p">(</span><span class="s2">&quot;YHR174W&quot;</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_30_0.png" /></p>
<h3 id="plot_signal">plot_signal</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">plot_signal</span><span class="p">)</span>
</pre></div>

<p><strong>plot_signal</strong> (ref_id, start, end, kind, split_samples, figsize, palette, plot_style)</p>
<p>Plot the dwell time and median intensity distribution position per position
Pointless for more than 50 positions at once as it becomes hard to distinguish</p>
<hr />
<ul>
<li><strong>ref_id</strong> (required) [str]</li>
</ul>
<p>Valid reference id name in the database</p>
<ul>
<li><strong>start</strong> (default: None) [int]</li>
</ul>
<p>Start coordinate</p>
<ul>
<li><strong>end</strong> (default: None) [int]</li>
</ul>
<p>End coordinate (included)</p>
<ul>
<li><strong>kind</strong> (default: violinplot) [str]</li>
</ul>
<p>Kind of plot {violinplot, boxenplot, swarmplot}</p>
<ul>
<li><strong>split_samples</strong> (default: False) [bool]</li>
</ul>
<p>If samples for a same condition are represented separatly. If false they are merged per condition</p>
<ul>
<li><strong>figsize</strong> (default: (30, 10)) [tuple]</li>
</ul>
<p>Length and heigh of the output plot</p>
<ul>
<li><strong>palette</strong> (default: Set2) [str]</li>
</ul>
<p>Colormap. See <a href="https://matplotlib.org/users/colormaps.html">https://matplotlib.org/users/colormaps.html</a>, <a href="https://matplotlib.org/examples/color/named_colors.html">https://matplotlib.org/examples/color/named_colors.html</a></p>
<ul>
<li><strong>plot_style</strong> (default: ggplot) [str]</li>
</ul>
<p>Matplotlib plotting style. See <a href="https://matplotlib.org/users/style_sheets.html">https://matplotlib.org/users/style_sheets.html</a></p>
<h4 id="examples-from-simulated-dataset_1">Examples from simulated dataset</h4>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_signal</span> <span class="p">(</span><span class="s2">&quot;ref_0000&quot;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">75</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_34_0.png" /></p>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_signal</span> <span class="p">(</span><span class="s2">&quot;ref_0001&quot;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">125</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;swarmplot&quot;</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_35_0.png" /></p>
<h4 id="example-from-real-yeast-dataset">Example from real yeast dataset</h4>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/yeast_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/yeast/Yeast_transcriptome.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_signal</span> <span class="p">(</span><span class="s2">&quot;YHR174W&quot;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">665</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">700</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;boxenplot&quot;</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_37_0.png" /></p>
<h3 id="plot_coverage">plot_coverage</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">plot_coverage</span><span class="p">)</span>
</pre></div>

<p><strong>plot_coverage</strong> (ref_id, start, end, scale, split_samples, figsize, palette, plot_style)</p>
<p>Plot the read coverage over a reference for all samples analysed</p>
<hr />
<ul>
<li><strong>ref_id</strong> (required) [str]</li>
</ul>
<p>Valid reference id name in the database</p>
<ul>
<li><strong>start</strong> (default: None) [int]</li>
</ul>
<p>Start coordinate</p>
<ul>
<li><strong>end</strong> (default: None) [int]</li>
</ul>
<p>End coordinate (included)</p>
<ul>
<li>
<p><strong>scale</strong> (default: False) [bool]</p>
</li>
<li>
<p><strong>split_samples</strong> (default: False) [bool]</p>
</li>
<li>
<p><strong>figsize</strong> (default: (30, 5)) [tuple]</p>
</li>
</ul>
<p>Length and heigh of the output plot</p>
<ul>
<li><strong>palette</strong> (default: Set2) [str]</li>
</ul>
<p>Colormap. See <a href="https://matplotlib.org/users/colormaps.html">https://matplotlib.org/users/colormaps.html</a>, <a href="https://matplotlib.org/examples/color/named_colors.html">https://matplotlib.org/examples/color/named_colors.html</a></p>
<ul>
<li><strong>plot_style</strong> (default: ggplot) [str]</li>
</ul>
<p>Matplotlib plotting style. See <a href="https://matplotlib.org/users/style_sheets.html">https://matplotlib.org/users/style_sheets.html</a></p>
<h4 id="example-from-real-yeast-dataset_1">Example from real yeast dataset</h4>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/yeast_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/yeast/Yeast_transcriptome.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_coverage</span> <span class="p">(</span><span class="s2">&quot;YHR174W&quot;</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_41_0.png" /></p>
<h3 id="plot_kmers_stats">plot_kmers_stats</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">plot_kmers_stats</span><span class="p">)</span>
</pre></div>

<p><strong>plot_kmers_stats</strong> (ref_id, start, end, split_samples, figsize, palette, plot_style)</p>
<p>Fancy version of <code class="codehilite"><span class="n">plot_coverage</span></code> that also report missing, mismatching and undefined kmers status from Nanopolish</p>
<hr />
<ul>
<li><strong>ref_id</strong> (required) [str]</li>
</ul>
<p>Valid reference id name in the database</p>
<ul>
<li><strong>start</strong> (default: None) [int]</li>
</ul>
<p>Start coordinate</p>
<ul>
<li><strong>end</strong> (default: None) [int]</li>
</ul>
<p>End coordinate (included)</p>
<ul>
<li>
<p><strong>split_samples</strong> (default: False) [bool]</p>
</li>
<li>
<p><strong>figsize</strong> (default: (30, 10)) [tuple]</p>
</li>
</ul>
<p>Length and heigh of the output plot</p>
<ul>
<li><strong>palette</strong> (default: Accent) [str]</li>
</ul>
<p>Colormap. See <a href="https://matplotlib.org/users/colormaps.html">https://matplotlib.org/users/colormaps.html</a>, <a href="https://matplotlib.org/examples/color/named_colors.html">https://matplotlib.org/examples/color/named_colors.html</a></p>
<ul>
<li><strong>plot_style</strong> (default: ggplot) [str]</li>
</ul>
<p>Matplotlib plotting style. See <a href="https://matplotlib.org/users/style_sheets.html">https://matplotlib.org/users/style_sheets.html</a></p>
<h4 id="example-from-real-yeast-dataset_2">Example from real yeast dataset</h4>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/yeast_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/yeast/Yeast_transcriptome.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_kmers_stats</span> <span class="p">(</span><span class="s2">&quot;YHR174W&quot;</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_45_0.png" /></p>
<h3 id="plot_position">plot_position</h3>
<div class="codehilite"><pre><span></span><span class="n">jhelp</span><span class="p">(</span><span class="n">SampCompDB</span><span class="o">.</span><span class="n">plot_position</span><span class="p">)</span>
</pre></div>

<p><strong>plot_position</strong> (ref_id, pos, split_samples, figsize, palette, plot_style, xlim, ylim, alpha, pointSize, scatter, kde, model, gmm_levels)</p>
<p>Plot the dwell time and median intensity at the given position as a scatter plot.</p>
<hr />
<ul>
<li><strong>ref_id</strong> (required) [str]</li>
</ul>
<p>Valid reference id name in the database</p>
<ul>
<li><strong>pos</strong> (default: None) [int]</li>
</ul>
<p>Position of interest</p>
<ul>
<li><strong>split_samples</strong> (default: False) [bool]</li>
</ul>
<p>If True, samples for a same condition are represented separately. If False, they are merged per condition</p>
<ul>
<li><strong>figsize</strong> (default: (30, 10)) [tuple]</li>
</ul>
<p>Length and heigh of the output plot</p>
<ul>
<li><strong>palette</strong> (default: Set2) [str]</li>
</ul>
<p>Colormap. See <a href="https://matplotlib.org/users/colormaps.html">https://matplotlib.org/users/colormaps.html</a>, <a href="https://matplotlib.org/examples/color/named_colors.html">https://matplotlib.org/examples/color/named_colors.html</a></p>
<ul>
<li><strong>plot_style</strong> (default: ggplot) [str]</li>
</ul>
<p>Matplotlib plotting style. See <a href="https://matplotlib.org/users/style_sheets.html">https://matplotlib.org/users/style_sheets.html</a></p>
<ul>
<li><strong>xlim</strong> (default: (None, None)) [tuple]</li>
</ul>
<p>A tuple of explicit limits for the x axis</p>
<ul>
<li><strong>ylim</strong> (default: (None, None)) [tuple]</li>
</ul>
<p>A tuple of explicit limits for the y axis</p>
<ul>
<li>
<p><strong>alpha</strong> (default: 0.3) [float]</p>
</li>
<li>
<p><strong>pointSize</strong> (default: 20) [int]</p>
</li>
</ul>
<p>int specifying the point size for the scatter plot</p>
<ul>
<li><strong>scatter</strong> (default: True) [bool]</li>
</ul>
<p>if True, plot the individual data points</p>
<ul>
<li><strong>kde</strong> (default: True) [bool]</li>
</ul>
<p>plot the KDE of the intensity/dwell bivarariate distributions in the two samples</p>
<ul>
<li><strong>model</strong> (default: False) [bool]</li>
</ul>
<p>If true, plot the GMM density estimate</p>
<ul>
<li><strong>gmm_levels</strong> (default: 50) [int]</li>
</ul>
<p>number of contour lines to use for the GMM countour plot</p>
<h4 id="example-from-simulated-dataset">Example from simulated dataset</h4>
<div class="codehilite"><pre><span></span><span class="c1"># Reload DB</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">SampCompDB</span> <span class="p">(</span><span class="n">db_fn</span> <span class="o">=</span> <span class="s2">&quot;results/simulated_SampComp.db&quot;</span><span class="p">,</span> <span class="n">fasta_fn</span> <span class="o">=</span> <span class="s2">&quot;references/simulated/ref.fa&quot;</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;warning&quot;</span><span class="p">)</span>
<span class="c1"># Plot</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">plot_position</span> <span class="p">(</span><span class="s2">&quot;ref_0000&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="o">=</span><span class="mi">82</span><span class="p">)</span>
</pre></div>

<p><img alt="png" src="output_49_0.png" /></p></article></div></div></main><footer class="md-footer"><div class="md-footer-nav"><nav class="md-footer-nav__inner md-grid"><a href="../SampComp_usage/" title="Compare samples with SampComp" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev"><div class="md-flex__cell md-flex__cell--shrink"><i class="md-icon md-icon--arrow-back md-footer-nav__button"></i></div><div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title"><span class="md-flex__ellipsis"><span class="md-footer-nav__direction">Previous</span>Compare samples with SampComp</span></div></a><a href="../SimReads_usage/" title="Generate simulated reads" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next"><div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title"><span class="md-flex__ellipsis"><span class="md-footer-nav__direction">Next</span>Generate simulated reads</span></div><div class="md-flex__cell md-flex__cell--shrink"><i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i></div></a></nav></div><div class="md-footer-meta md-typeset"><div class="md-footer-meta__inner md-grid"><div class="md-footer-copyright"><div class="md-footer-copyright__highlight">Copyright &copy; 2019 Adrien Leger & Tommaso Leonardi</div>powered by <a href="https://www.mkdocs.org">MkDocs</a> and <a href="https://squidfunk.github.io/mkdocs-material/">Material for MkDocs</a></div></div></div></footer></div><script src="../../assets/javascripts/application.b260a35d.js"></script><script>app.initialize({version:"1.0.4",url:{base:"../.."}})</script></body></html>