# How to use Nanocompore

Nanocompore was designed to be used either through a python API or a command line interface. The package contains 3 main modules: `SampComp`, `SampCompDB` and `SimReads`

### SampComp

This is the main module which compares the signal of 2 experimental conditions. It takes Nanopolished datasets as input and generates a database containing all the results as output. `SampComp` has to be run first either through the API (`nanocompore.SampComp.SampComp`) or the CLI (`nanocompore sampcomp`).

* [SampComp Usage](https://github.com/tleonardi/nanocompore/blob/master/docs/demo/SampComp_usage.ipynb)


### SampCompDB

`SampCompDB` is a wrapper around the DBM object database generated by `SampComp`. This module performs secondary statistical analyses and provide simple high level functions to plot, explore and export the results. At the moment `SampCompDB` is only accessible through the interactive python API (`nanocompore.SampCompDB.SampCompDB`). We strongly recommend to use [jupyter notebook](https://jupyter.org/).

* [SampCompDB Usage](https://github.com/tleonardi/nanocompore/blob/master/docs/demo/SampComp_usage.ipynb)

### SimReads

This module can be used to generate artificial datasets based on a model file obtained from IVT generated RNA sequenced by direct RNA sequencing ([Datasets](https://github.com/nanopore-wgs-consortium/NA12878/blob/master/nanopore-human-transcriptome/fastq_fast5_bulk.md), from the Nanopore RNA consortium). In addition, one can also simulate the presence of modifications by allowing to deviate from the model for selected positions.

* [Simulate_reads Usage](https://github.com/tleonardi/nanocompore/blob/master/docs/demo/SimReads_usage.ipynb)
